Api for product

Ayrum1158 opened the pull request.

Commit:
Added CategoryId property to ProductEntity
Added OneToMany relationship between ProductEntity and CategoryEntity

Signed-off-by: Oleksandr Korotetskyi <o.korotetskyi@chisw.com>

Commit:
Added ProductsController skeleton
Added ProductViewModel

Signed-off-by: Oleksandr Korotetskyi <o.korotetskyi@chisw.com>

Commit:
WIP
Renamed "Category" service and interface to "Categories"
Created ProductsService and IProductsService

Signed-off-by: Oleksandr Korotetskyi <o.korotetskyi@chisw.com>

Commit:
Merge branch 'CRUDForCategory' into APIForProduct

Signed-off-by: Oleksandr Korotetskyi <o.korotetskyi@chisw.com>

Commit:
Fixed after-merge namespace naming

Signed-off-by: Oleksandr Korotetskyi <o.korotetskyi@chisw.com>

Commit:
Changed seeding method to seed entities with positive Ids

Signed-off-by: Oleksandr Korotetskyi <o.korotetskyi@chisw.com>

Commit:
Made GenericRepository and it's interface asynchronous

Signed-off-by: Oleksandr Korotetskyi <o.korotetskyi@chisw.com>

Commit:
Made all ProductsService methods async
Fixed typos
Worked on GenericRepository asynchrony further
Changed some ProductsService  methods to return created entities
Fixed ProductsService AddCategory to always return Id: 0
Renamed ResultContract to ServiceResult
Made working Products API and service
Fixed possibility to update category setting existing name

Known issues: ProductsController POST throws object disposed exception

Signed-off-by: Oleksandr Korotetskyi <o.korotetskyi@chisw.com>

Commit:
Fixed wrong usage of IsPresentInDb method
Fixed FirstOrDefaultAsync method

Signed-off-by: Oleksandr Korotetskyi <o.korotetskyi@chisw.com>

Commit:
Fixed wrong id comparison in PUT action of productsController
Created ProductRepository

Known issues: no update on product update

Signed-off-by: Oleksandr Korotetskyi <o.korotetskyi@chisw.com>

Commit:
Fixed bug of not updating product in ProductsRepository

Signed-off-by: Oleksandr Korotetskyi <o.korotetskyi@chisw.com>

Commit:
Added action, service method and repository metrod
for getting paginated and specifically sorted products

Signed-off-by: Oleksandr Korotetskyi <o.korotetskyi@chisw.com>

solianyk-sergey commented on InternetStoreCHIReloaded/API/Controllers/ProductsController.cs:
```diff
@@ -0,0 +1,94 @@
+        }
+
+        [HttpGet]
+        public async Task<GenericResponse<IEnumerable<ProductViewModel>>> Get()// return all, add pagination next
```
It's helpful for you to leave those comments now.
In live projects such simple comments are redundant because REST principals are used and method responsibility is obvious.
Swagger description could be used for such type of method description.

solianyk-sergey replied:
It's helpful for you to leave those comments now.
In live projects such simple comments are redundant because REST principals are used and method responsibility is obvious.
Swagger description could be used for such type of method description.

solianyk-sergey commented on InternetStoreCHIReloaded/API/Controllers/ProductsController.cs:
```diff
@@ -0,0 +1,94 @@
+        }
+
+        [HttpGet("{pageSize}/{page}/{filterString}")]// filter should be json object corresponding to ProductsFilterViewModel
+        public async Task<GenericResponse<List<ProductViewModel>>> Get(int pageSize, int page, string filterString)
```
It's better to use POST to pass filter and pagination data

solianyk-sergey replied:
Request body could be mapped to some filtration model

solianyk-sergey commented on InternetStoreCHIReloaded/BLL/Services/ProductsService.cs:
```diff
@@ -0,0 +1,219 @@
+                p => prop.GetValue(p, null),
+                filter.SortDirection,
+                pageSize,
+                page);
```
All filtration logic should be part of the repository.

solianyk-sergey commented on InternetStoreCHIReloaded/DAL/Extensions/ModelBuilderExtentions.cs:
```diff
@@ -46,7 +46,8 @@ public static void Seed(this ModelBuilder modelBuilder)
                     Price = (decimal)1337.69,
                     CreatedDate = timeNow,
-                    UpdatedDate = timeNow
+                    UpdatedDate = timeNow,
```
Let's hardcode the date instead of using DateTime.UtcNow. Otherwise, default data will be updated for each migration

solianyk-sergey commented on InternetStoreCHIReloaded/DAL/Repositories/ProductsRepository.cs:
```diff
@@ -0,0 +1,59 @@
+        {
+            IEnumerable<ProductEntity> products = null;
+
+            var findTask = Task.Run(() =>
```
Adding Task.Run() is redundant. Instead, result operation should get all data async from the database by using .ToListAsync()

solianyk-sergey commented on InternetStoreCHIReloaded/DAL/Repositories/ProductsRepository.cs:
```diff
@@ -0,0 +1,59 @@
+                }
+
+                int from = pageSize * (page - 1);
+                products = products.Skip(from).Take(pageSize).ToList();
```
Please, change to `.ToListAsync()`

solianyk-sergey commented on InternetStoreCHIReloaded/BLL/Extentions/ExpressionExtentions.cs:
```diff
@@ -0,0 +1,58 @@
+
+namespace BLL.Extentions
+{
+    public static class ExpressionExtentions
```
Do we need this extensions at all?

Ayrum1158 replied:
Yes. Only one of thease methods is actually used, but combining expressions without them is hell.

solianyk-sergey commented on InternetStoreCHIReloaded/BLL/Services/ProductsService.cs:
```diff
@@ -0,0 +1,219 @@
+                whereExpr = whereExpr.And((pe) => pe.Price >= filter.FromPrice);
+
+            if (filter.ToPrice != null)
+                whereExpr = whereExpr.And((pe) => pe.Price <= filter.ToPrice);
```
This logic is overcomplicated. 
We could store IQueryable variable like:
```
var productsQuery = _context.Products.AsQueryable();
```

Later we could add as many .Where() conditions as you wish, like:

```
if (minPrice != null)
{
  productsQuery = productsQuery.Where(i => i.Price >= minPrice.Value);
}
```

This code should be part of the DAL repository method.

solianyk-sergey reviewed (commented):
Please check comments

Commit:
Removed redundant usage of Task.Run()
in ProductsRepository.FindSortAndPaginateAll method
Moved product filtering code from service to repository
Removed Expression class usage from product filtering

Signed-off-by: Oleksandr Korotetskyi <o.korotetskyi@chisw.com>

Commit:
Changed default data seeding date to hardcoded value
Added related migration

Signed-off-by: Oleksandr Korotetskyi <o.korotetskyi@chisw.com>

Commit:
Fixed exception when sorting products results
Changed Get action for sorted products to Post

Signed-off-by: Oleksandr Korotetskyi <o.korotetskyi@chisw.com>

Commit:
Removed redundant comments in controllers

Signed-off-by: Oleksandr Korotetskyi <o.korotetskyi@chisw.com>

Commit:
Made products get query from db truly async
Started using ThenBy for query consistency

Signed-off-by: Oleksandr Korotetskyi <o.korotetskyi@chisw.com>

Commit:
Refactored GetProperty method in ProductsRepository

Signed-off-by: Oleksandr Korotetskyi <o.korotetskyi@chisw.com>

Commit:
Products filtered query is now fully executing on Database side

Signed-off-by: Oleksandr Korotetskyi <o.korotetskyi@chisw.com>
